### PR TITLE
[FLINK-19022][runtime]Register the TerminationFuture of ResourceManager and Dispatcher with DispatcherResourceManagerComponent

### PR DESCRIPTION
**Pull request by Tartarus0zm**
## What is the purpose of the change

Help analyze the cause of the abnormality and exit the program normally

## Brief change log

1. add log when RpcEndpoint start/stop failed.
2.Register the TerminationFuture of ResourceManager and Dispatcher with DispatcherResourceManagerComponent.

## Verifying this change

no

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by Tartarus0zm** [comment]
@tillrohrmann  please take a look, if you have time, thanks

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java [hunk 1]
```diff
@@ -88,6 +95,18 @@
...
+			if (throwable != null) {
+				shutDownFuture.completeExceptionally(throwable);
+			} else {
+				shutDownFuture.complete(ApplicationStatus.SUCCEEDED);
+			}
+			if (isRunning.get()) {
+				fatalErrorHandler.onFatalError(throwable);
+			}
+		};
+		dispatcherRunner.getTerminationFuture().whenComplete(terminateAction);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:108]: I think the `DispatcherRunner.getTerminationFuture` won't complete if the `Dispatcher` terminates. The problem is that we only forward the result of the `DispatcherLeaderProcess.terminationFuture` to the `DispatcherRunner.terminationFuture` if we call `closeAsync`. I would suggest to add tests to verify the intended behavior.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java [hunk 2]
```diff
@@ -107,13 +112,18 @@ public void grantLeadership(UUID leaderSessionID) {
...
 
-		final DispatcherLeaderProcess newDispatcherLeaderProcess = dispatcherLeaderProcess;
-		FutureUtils.assertNoException(
-			previousDispatcherLeaderProcessTerminationFuture.thenRun(newDispatcherLeaderProcess::start));
+			final DispatcherLeaderProcess newDispatcherLeaderProcess = dispatcherLeaderProcess;
+			FutureUtils.assertNoException(
+				previousDispatcherLeaderProcessTerminationFuture.thenRun(newDispatcherLeaderProcess::start));
+		} catch (Throwable t) {
+			terminationFuture.completeExceptionally(t);
+			throw t;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java:125]: Why is this needed?

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @Tartarus0zm. I think the current implementation still has a couple of problems. The main part is the forwarding and usage of the termination future of the `DispatcherRunner`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java (hunk 1 above)
Comment (Tartarus0zm) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:108]: But how we get the Dispatcher's TerminationFuture in DispatcherResourceManagerComponent?
Please give a hint, thanks

Comment (Tartarus0zm) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:108]: @tillrohrmann  I need some help,  I took a look at this piece of code,  `DispatcherRunner.terminationFuture` is not connected to `Dispatcher`, so how we get the `Dispatcher.TerminationFuture` in `DispatcherResourceManagerComponent`?
Please give a hint, thanks

Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:108]: @tillrohrmann is on vacation for a few weeks starting from today. He won't be able to resolve your questions.

I would suggest to follow his suggestions and add test cases reproducing the problem first. This should make it easier finding the right location in the code in the code handle errors.

Comment (Tartarus0zm) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:108]: @rmetzger  thanks, I try to solve it myself

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java [hunk 3]
```diff
@@ -74,22 +79,46 @@
...
+		failOnPrematureTermination(dispatcherGatewayCompletableFuture);
 	}
 
 	private void registerShutDownFuture() {
 		FutureUtils.forward(dispatcherRunner.getShutDownFuture(), shutDownFuture);
 	}
 
+	private void failOnPrematureTermination(CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
+		dispatcherGatewayCompletableFuture.whenComplete((dispatcher, throwable) -> {
+			if (dispatcher != null && dispatcher instanceof Dispatcher) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:104]: This breaks the DispatcherGateway abstraction and reinforces the assumption that the RM and Dispatcher run in the same JVM which may change in the future.
Why can we not use the shutdown future of the `DispatcherRunner`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java [hunk 4]
```diff
@@ -74,22 +79,46 @@
...
+				CompletableFuture.anyOf(((Dispatcher) dispatcher).getTerminationFuture(), resourceManager.getTerminationFuture())
+					.whenComplete((ignored, t) -> {
+						if (isRunning.get()) {
+							fatalErrorHandler.onFatalError(t);
+						}
+					});
+			} else {
+				LOG.warn("DispatcherGateway({}) not instance of Dispatcher or failed to get DispatcherGateway!", dispatcher, throwable);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:112]: This kind of implementation details should not be written to the logs.

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java [hunk 5]
```diff
@@ -74,22 +79,46 @@
...
 		FutureUtils.forward(dispatcherRunner.getShutDownFuture(), shutDownFuture);
 	}
 
+	private void failOnPrematureTermination(CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
+		dispatcherGatewayCompletableFuture.whenComplete((dispatcher, throwable) -> {
+			if (dispatcher != null && dispatcher instanceof Dispatcher) {
+				CompletableFuture.anyOf(((Dispatcher) dispatcher).getTerminationFuture(), resourceManager.getTerminationFuture())
+					.whenComplete((ignored, t) -> {
+						if (isRunning.get()) {
+							fatalErrorHandler.onFatalError(t);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:108]: It would be good to log a warning here that the DRMC shut down because the dispatcher/rm unexpected terminated.

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java [hunk 6]
```diff
@@ -74,22 +79,46 @@
...
 
 	private void registerShutDownFuture() {
 		FutureUtils.forward(dispatcherRunner.getShutDownFuture(), shutDownFuture);
 	}
 
+	private void failOnPrematureTermination(CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
+		dispatcherGatewayCompletableFuture.whenComplete((dispatcher, throwable) -> {
+			if (dispatcher != null && dispatcher instanceof Dispatcher) {
+				CompletableFuture.anyOf(((Dispatcher) dispatcher).getTerminationFuture(), resourceManager.getTerminationFuture())
+					.whenComplete((ignored, t) -> {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:106]: It may be a good idea to check whether t is indeed not null and log something accordingly if this is not the case. If it is null, then the process shut down without an error unexpectedly, which if I understand things correctly should not happen.

**Review by zentol — CHANGES_REQUESTED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java (hunk 3 above)
Comment (Tartarus0zm) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:104]: The shutdown future of the `DispatcherRunner` has been registed with the shutDown Future of `DispatcherResourceManagerComponent`.  I think there should be no need to deal with it again.
How about that only deals with the Termination Future of ResourceManager here?

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java (hunk 6 above)
Comment (Tartarus0zm) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:106]: We has modify `Dispatcher` and `ResourceManager` catch Exception to catch Throwable ，and then call `onFatalError` , so I think `DispatcherResourceManagerComponent#failOnPrematureTermination` is not very necessary, but @tillrohrmann  thinks it is safer

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:106]: I don't see how that relates to my comment. We will call the fatal error handler either way; the issue is that if t == null no reason for the shutdown will be logged-

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java (hunk 3 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:104]: Are the DispatcherRunner and Dispatcher termination futures linked in any way? As in, if the Dispatcher shuts down _at any point_, then it is guaranteed that the DispatcherRunner termination future is also being completed?

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java (hunk 6 above)
Comment (Tartarus0zm) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:106]: sorry, my understanding is wrong.

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java (hunk 3 above)
Comment (Tartarus0zm) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:104]: if we use the shutdown future of the DispatcherRunner like this
```
private void failOnPrematureTermination() {
	CompletableFuture.anyOf(dispatcherRunner.getShutDownFuture(), resourceManager.getTerminationFuture())
		.whenComplete((ignored, t) -> {
			if (t == null) {
				LOG.error("Dispatcher/ResourceManager shut down because something unexpected happen!");
			}
			if (isRunning.get()) {
				LOG.warn("DispatcherResourceManagerComponent shut down because the Dispatcher/ResourceManager unexpected terminated.");
				fatalErrorHandler.onFatalError(t);
			}
		});
}
```
in per-job mode, `ClusterEntrypoint.shutDownAsync()` may not have a chance to execute, because `ClusterEntrypoint.onFatalError` will call `System.exit(RUNTIME_FAILURE_RETURN_CODE);`
Do you have any good suggestions for revision?

Comment (Tartarus0zm) [reply flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java:104]: We first execute `failOnPrematureTermination();` and then execute `registerShutDownFuture();` 
How about this?  @zentol

**Comment by tillrohrmann** [comment]
Sorry for the late response @Tartarus0zm. I have take a look at your solution and reworked it a bit. Please take a look at the pushed commits.

@zentol I also appreciate your review of the changes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java [hunk 7]
```diff
@@ -139,6 +139,8 @@ public final UUID getLeaderSessionId() {
 	private void closeInternal() {
 		log.info("Stopping {}.", getClass().getSimpleName());
 
+		state = State.STOPPED;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java:143]: Does this change belong into `cd14c5907bdbd88861f9e74448ded52d5c15a7a6`? (Because it requires other components to be stopped after the leader process transitioned to STOPPED.

File: flink-runtime/src/test/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponentTest.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	private DispatcherResourceManagerComponent createDispatcherResourceManagerComponent(
+			TestingFatalErrorHandler fatalErrorHandler,
+			TestingResourceManagerService resourceManagerService) {
+		return new DispatcherResourceManagerComponent(
+			TestingDispatcherRunner.newBuilder().build(),
+				resourceManagerService,
+			new SettableLeaderRetrievalService(),
+			new SettableLeaderRetrievalService(),
+				FutureUtils::completedVoidFuture,
+				fatalErrorHandler);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponentTest.java:70]: odd indentation

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java [hunk 9]
```diff
@@ -144,6 +148,58 @@ public void testGenerateTaskManagerResourceIDWithLocalRpcService() throws Except
...
+				(configuration,
+				resourceID,
+				rpcService,
+				highAvailabilityServices,
+				heartbeatServices,
+				metricRegistry,
+				blobCacheService,
+				localCommunicationOnly,
+				externalResourceInfoProvider,
+				fatalErrorHandler) -> taskExecutorService);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:168]: maybe introduce a utility method that returns this factory; it is easy to miss that the list of arguments is part of the factory, and not the constructor.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java [hunk 10]
```diff
@@ -144,6 +148,58 @@ public void testGenerateTaskManagerResourceIDWithLocalRpcService() throws Except
...
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:194]: same as above

**Review by zentol — APPROVED** [review]
+1

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java (hunk 7 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/AbstractDispatcherLeaderProcess.java:143]: Yes I think it is correct. `State.STOPPED` effectively means that `closeInternal` has been called. But it does not say that the stopping is complete.

File: flink-runtime/src/test/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponentTest.java (hunk 8 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponentTest.java:70]: You are right. Somehow IntelliJ uses a different indentation nowadays... Should check the configuration.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java (hunk 9 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:168]: Good point. I will update it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java (hunk 10 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunnerTest.java:194]: Good idea. Will do.

**Comment by tillrohrmann** [comment]
Thanks for the review @zentol. Merging this PR now.
